File: packages/verify/pq-hybrid.js
/**
 * @emilia-protocol/verify: HYBRID classical + post-quantum signature envelope
 * (EP-HYBRID-v1) for EP INFRASTRUCTURE keys (transparency-log, directory,
 * checkpoint signing keys).
 *
 * See docs/POST-QUANTUM-MIGRATION.md (section "EP-HYBRID-v1 hybrid envelope").
 *
 * ENVELOPE
 *   {
 *     alg: 'EP-HYBRID-v1',
 *     signature_algos: ['Ed25519', 'ML-DSA-65'],   // canonical order, exact set
 *     sigs: {
 *       'Ed25519':   '<base64url>',
 *       'ML-DSA-65': '<base64url>'
 *     }
 *   }
 *
 * ANTI-STRIPPING (the property this module exists for)
 *   BOTH signatures are computed over a domain-separated signing input that
 *   INCLUDES a canonical encoding of `signature_algos`:
 *
 *     signing_input = UTF8('emilia-protocol/pq-hybrid/v1') || 0x00
 *                  || UTF8(JSON.stringify(signature_algos)) || 0x00
 *                  || message
 *
 *   Therefore an attacker who strips the ML-DSA-65 signature and presents the
 *   Ed25519 signature as if it were a plain / classical-only signature fails:
 *   the Ed25519 signature does not verify over the bare message, nor over a
 *   signing input that commits to a reduced algorithm set. Removing an
 *   algorithm from the set changes what was signed, so every remaining
 *   signature fails. verifyHybrid() additionally requires the PRESENTED
 *   `signature_algos` to equal the registered EP-HYBRID-v1 set exactly
 *   (order-sensitive) and requires one signature per committed algorithm,
 *   no more, no fewer.
 *
 * FAIL-CLOSED
 *   - Missing / malformed envelope, message, or key material refuses.
 *   - Tampered or reduced `signature_algos` refuses (algo_set_mismatch).
 *   - A committed algorithm with no signature refuses (missing_signature).
 *   - Either signature invalid refuses (classical_signature_invalid /
 *     pq_signature_invalid).
 *   - If no ML-DSA backend is available, verifyHybrid REFUSES with
 *     reason 'pq_backend_unavailable'. It NEVER skips the PQ leg and never
 *     returns verified:true on the classical leg alone.
 *
 * ML-DSA BACKEND (honesty)
 *   packages/verify ships with ZERO runtime dependencies, so this module does
 *   not (and must not) import an ML-DSA implementation statically. The ML-DSA
 *   leg is cryptographically live only when a backend is present, via either:
 *     1. caller injection: pass { mldsaBackend } implementing
 *        sign(messageBytes, secretKeyBytes) -> Uint8Array and
 *        verify(signatureBytes, messageBytes, publicKeyBytes) -> boolean; or
 *     2. lazy dynamic import of '@noble/post-quantum/ml-dsa.js' (ml_dsa65),
 *        which resolves only if that package is installed by the consumer
 *        (it is a devDependency at this repo's root for tests, NOT a
 *        dependency of this package).
 *   If neither is available the module refuses; absence of the backend is a
 *   refusal, never a crash and never a pass. If a caller injects a bogus
 *   always-true backend, the PQ leg is only as strong as that backend (i.e.
 *   vacuous); the classical leg and the algo-set commitment are still
 *   enforced independently, but the caller owns the PQ leg's honesty.
 *
 *   @noble/post-quantum v0.6.1 is a pure-JS implementation of FIPS 204
 *   ML-DSA. Per its own README it is self-audited by its authors and has NOT
 *   been independently audited; it is not a FIPS-validated module. This envelope does NOT make EP
 *   post-quantum secure by itself; see docs/POST-QUANTUM-MIGRATION.md for the
 *   migration phases and the pre-CRQC historical re-anchoring requirement.
 *
 * KEY / SIGNATURE ENCODING (matches index.js conventions)
 *   - Ed25519 public key: base64url SPKI DER (or a node crypto KeyObject).
 *   - Ed25519 private key: node crypto KeyObject (from generateKeyPairSync /
 *     createPrivateKey).
 *   - ML-DSA-65 public key: raw bytes (Uint8Array, 1952 bytes) or base64url.
 *   - ML-DSA-65 secret key: raw bytes (Uint8Array, 4032 bytes).
 *   - Signatures in the envelope: base64url strings.
 *
 * @license Apache-2.0
 */

import crypto from 'crypto';

export const HYBRID_ALG = 'EP-HYBRID-v1';

/**
 * The registered algorithm set for EP-HYBRID-v1, in canonical order.
 * v1 is a FIXED two-algorithm hybrid: exactly Ed25519 + ML-DSA-65.
 */
export const HYBRID_SIGNATURE_ALGOS = Object.freeze(['Ed25519', 'ML-DSA-65']);

/**
 * Domain-separation label. Trailing 0x00 separators in the signing input keep
 * label / algo-set / message unambiguous (JSON never contains a raw 0x00).
 */
export const HYBRID_DOMAIN = 'emilia-protocol/pq-hybrid/v1';

const REASONS = Object.freeze({
  INVALID_INPUT: 'invalid_input',
  INVALID_ENVELOPE: 'invalid_envelope',
  ALGO_SET_MISMATCH: 'algo_set_mismatch',
  MISSING_SIGNATURE: 'missing_signature',
  MISSING_KEY: 'missing_key',
  CLASSICAL_INVALID: 'classical_signature_invalid',
  PQ_INVALID: 'pq_signature_invalid',
  PQ_BACKEND_UNAVAILABLE: 'pq_backend_unavailable',
});
export { REASONS as HYBRID_REASONS };

// ---------------------------------------------------------------------------
// Signing input (the anti-stripping commitment)
// ---------------------------------------------------------------------------

/**
 * Build the domain-separated signing input BOTH legs sign.
 * Commits to the full signature_algos array; changing the array in any way
 * (strip, reorder, substitute) changes these bytes.
 *
 * @param {Uint8Array} messageBytes
 * @param {string[]} signatureAlgos
 * @returns {Buffer}
 */
export function hybridSigningInput(messageBytes, signatureAlgos) {
  if (!(messageBytes instanceof Uint8Array)) {
    throw new TypeError('hybridSigningInput: messageBytes must be a Uint8Array');
  }
  if (!Array.isArray(signatureAlgos) || signatureAlgos.length === 0
      || !signatureAlgos.every((a) => typeof a === 'string' && a.length > 0)) {
    throw new TypeError('hybridSigningInput: signatureAlgos must be a non-empty string array');
  }
  const zero = Buffer.from([0x00]);
  return Buffer.concat([
    Buffer.from(HYBRID_DOMAIN, 'utf8'), zero,
    Buffer.from(JSON.stringify(signatureAlgos), 'utf8'), zero,
    Buffer.from(messageBytes),
  ]);
}

// ---------------------------------------------------------------------------
// ML-DSA backend resolution (lazy, fail-closed)
// ---------------------------------------------------------------------------

/**
 * Try to load the default ML-DSA-65 backend (@noble/post-quantum). Returns a
 * backend object or null. NEVER throws; absence yields null so callers refuse.
 *
 * @returns {Promise<{sign: Function, verify: Function}|null>}
 */
export async function loadDefaultMldsaBackend() {
  try {
    const mod = await import('@noble/post-quantum/ml-dsa.js');
    const impl = mod.ml_dsa65;
    if (!impl || typeof impl.sign !== 'function' || typeof impl.verify !== 'function') return null;
    return {
      // noble API: sign(msg, secretKey) -> sig; verify(sig, msg, publicKey) -> boolean
      sign: (messageBytes, secretKeyBytes) => impl.sign(messageBytes, secretKeyBytes),
      verify: (signatureBytes, messageBytes, publicKeyBytes) => {
        try {
          return impl.verify(signatureBytes, messageBytes, publicKeyBytes) === true;
        } catch {
          return false; // malformed sig/key lengths refuse, never throw upward
        }
      },
    };
  } catch {
    return null;
  }
}

async function resolveBackend(mldsaBackend, mldsaBackendLoader) {
  if (mldsaBackend !== undefined && mldsaBackend !== null) {
    if (typeof mldsaBackend.verify !== 'function') return null; // malformed injection refuses
    return mldsaBackend;
  }
  const loader = typeof mldsaBackendLoader === 'function' ? mldsaBackendLoader : loadDefaultMldsaBackend;
  try {
    const b = await loader();
    if (!b || typeof b.verify !== 'function') return null;
    return b;
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Input normalization helpers
// ---------------------------------------------------------------------------

function toMessageBytes(message) {
  if (message instanceof Uint8Array) return message;
  if (typeof message === 'string') return Buffer.from(message, 'utf8');
  return null;
}

function toEd25519PublicKeyObject(key) {
  try {
    if (key && typeof key === 'object' && key.type === 'public') return key; // KeyObject
    if (typeof key === 'string' && key.length > 0) {
      const der = Buffer.from(key, 'base64url');
      return crypto.createPublicKey({ key: der, format: 'der', type: 'spki' });
    }
  } catch {
    return null;
  }
  return null;
}

function toRawBytes(key) {
  if (key instanceof Uint8Array) return key;
  if (typeof key === 'string' && key.length > 0) {
    try { return Buffer.from(key, 'base64url'); } catch { return null; }
  }
  return null;
}

function algosEqual(a, b) {
  return Array.isArray(a) && Array.isArray(b) && a.length === b.length
    && a.every((v, i) => v === b[i]);
}

// ---------------------------------------------------------------------------
// signHybrid
// ---------------------------------------------------------------------------

/**
 * Produce an EP-HYBRID-v1 envelope: Ed25519 AND ML-DSA-65 signatures, both
 * over the domain-separated signing input that commits to the full algo set.
 *
 * THROWS (fail-closed) if the ML-DSA backend is unavailable: this function
 * never silently emits a classical-only envelope.
 *
 * @param {Uint8Array|string} message
 * @param {object} keys
 * @param {import('crypto').KeyObject} keys.ed25519PrivateKey
 * @param {Uint8Array} keys.mldsaSecretKey - raw ML-DSA-65 secret key bytes
 * @param {object} [options]
 * @param {{sign:Function,verify:Function}} [options.mldsaBackend] - injected backend
 * @param {Function} [options.mldsaBackendLoader] - async () => backend|null
 * @returns {Promise<{alg:string, signature_algos:string[], sigs:Record<string,string>}>}
 */
export async function signHybrid(message, keys, options = {}) {
  const messageBytes = toMessageBytes(message);
  if (!messageBytes) throw new TypeError('signHybrid: message must be a Uint8Array or string');
  if (!keys || !keys.ed25519PrivateKey) throw new TypeError('signHybrid: keys.ed25519PrivateKey is required');
  const mldsaSecretKey = toRawBytes(keys.mldsaSecretKey);
  if (!mldsaSecretKey) throw new TypeError('signHybrid: keys.mldsaSecretKey is required');

  const backend = await resolveBackend(options.mldsaBackend, options.mldsaBackendLoader);
  if (!backend || typeof backend.sign !== 'function') {
    // Fail closed: never emit an envelope missing the PQ leg.
    throw new Error(`signHybrid: refusing to sign: ${REASONS.PQ_BACKEND_UNAVAILABLE}`);
  }

  const signatureAlgos = [...HYBRID_SIGNATURE_ALGOS];
  const signingInput = hybridSigningInput(messageBytes, signatureAlgos);

  const edSig = crypto.sign(null, signingInput, keys.ed25519PrivateKey);
  const pqSig = backend.sign(new Uint8Array(signingInput), mldsaSecretKey);
  if (!(pqSig instanceof Uint8Array) || pqSig.length === 0) {
    throw new Error('signHybrid: ML-DSA backend returned an invalid signature');
  }

  return {
    alg: HYBRID_ALG,
    signature_algos: signatureAlgos,
    sigs: {
      'Ed25519': Buffer.from(edSig).toString('base64url'),
      'ML-DSA-65': Buffer.from(pqSig).toString('base64url'),
    },
  };
}

// ---------------------------------------------------------------------------
// verifyHybrid
// ---------------------------------------------------------------------------

/**
 * Verify an EP-HYBRID-v1 envelope. verified:true requires ALL of:
 *   - well-formed envelope with alg 'EP-HYBRID-v1'
 *   - presented signature_algos EXACTLY equals the registered set
 *     ['Ed25519','ML-DSA-65'] (order-sensitive; this is also what both
 *     signatures commit to, so tampering fails twice)
 *   - exactly one signature per committed algorithm (no extras, none missing)
 *   - Ed25519 signature valid over the committed signing input
 *   - ML-DSA-65 signature valid over the committed signing input, checked by
 *     a REAL backend; if no backend is available the result is
 *     { verified:false, reason:'pq_backend_unavailable' } - the PQ leg is
 *     never skipped.
 *
 * @param {Uint8Array|string} message
 * @param {object} envelope - { alg, signature_algos, sigs }
 * @param {object} keys
 * @param {string|import('crypto').KeyObject} keys.ed25519PublicKey - base64url SPKI DER or KeyObject
 * @param {Uint8Array|string} keys.mldsaPublicKey - raw ML-DSA-65 public key bytes or base64url
 * @param {object} [options]
 * @param {{verify:Function}} [options.mldsaBackend] - injected backend
 * @param {Function} [options.mldsaBackendLoader] - async () => backend|null
 * @returns {Promise<{verified:boolean, reason:string|null, checks:object}>}
 */
export async function verifyHybrid(message, envelope, keys, options = {}) {
  /** @type {{envelope: boolean, algo_set: boolean, classical_signature: boolean|null, pq_signature: boolean|null}} */
  const checks = { envelope: false, algo_set: false, classical_signature: null, pq_signature: null };
  const refuse = (reason) => ({ verified: false, reason, checks });

  const messageBytes = toMessageBytes(message);
  if (!messageBytes) return refuse(REASONS.INVALID_INPUT);

  // 1. Envelope shape (fail closed on anything unexpected)
  if (!envelope || typeof envelope !== 'object' || Array.isArray(envelope)) return refuse(REASONS.INVALID_ENVELOPE);
  if (envelope.alg !== HYBRID_ALG) return refuse(REASONS.INVALID_ENVELOPE);
  if (!envelope.sigs || typeof envelope.sigs !== 'object' || Array.isArray(envelope.sigs)) return refuse(REASONS.INVALID_ENVELOPE);
  checks.envelope = true;

  // 2. Algo-set commitment: presented set must EXACTLY equal the registered
  //    set. (Both signatures also commit to it cryptographically below.)
  if (!algosEqual(envelope.signature_algos, HYBRID_SIGNATURE_ALGOS)) {
    return refuse(REASONS.ALGO_SET_MISMATCH);
  }
  checks.algo_set = true;

  // 3. Exactly one signature per committed algorithm; extras refuse.
  const sigKeys = Object.keys(envelope.sigs);
  if (sigKeys.length !== HYBRID_SIGNATURE_ALGOS.length) {
    return refuse(sigKeys.length < HYBRID_SIGNATURE_ALGOS.length ? REASONS.MISSING_SIGNATURE : REASONS.INVALID_ENVELOPE);
  }
  for (const algo of HYBRID_SIGNATURE_ALGOS) {
    if (typeof envelope.sigs[algo] !== 'string' || envelope.sigs[algo].length === 0) {
      return refuse(REASONS.MISSING_SIGNATURE);
    }
  }

  // 4. Key material (fail closed on missing/invalid input)
  if (!keys || !keys.ed25519PublicKey || !keys.mldsaPublicKey) return refuse(REASONS.MISSING_KEY);
  const edKey = toEd25519PublicKeyObject(keys.ed25519PublicKey);
  if (!edKey) return refuse(REASONS.MISSING_KEY);
  const pqKey = toRawBytes(keys.mldsaPublicKey);
  if (!pqKey || pqKey.length === 0) return refuse(REASONS.MISSING_KEY);

  const signingInput = hybridSigningInput(messageBytes, envelope.signature_algos);

  // 5. Classical leg (Ed25519) over the committed signing input
  let edOk = false;
  try {
    edOk = crypto.verify(null, signingInput, edKey, Buffer.from(envelope.sigs['Ed25519'], 'base64url'));
  } catch {
    edOk = false;
  }
  checks.classical_signature = edOk === true;
  if (!checks.classical_signature) return refuse(REASONS.CLASSICAL_INVALID);

  // 6. PQ leg (ML-DSA-65). No backend => REFUSE. Never skip, never pass.
  const backend = await resolveBackend(options.mldsaBackend, options.mldsaBackendLoader);
  if (!backend) return refuse(REASONS.PQ_BACKEND_UNAVAILABLE);
  let pqOk = false;
  try {
    pqOk = backend.verify(
      new Uint8Array(Buffer.from(envelope.sigs['ML-DSA-65'], 'base64url')),
      new Uint8Array(signingInput),
      new Uint8Array(pqKey),
    ) === true;
  } catch {
    pqOk = false;
  }
  checks.pq_signature = pqOk;
  if (!pqOk) return refuse(REASONS.PQ_INVALID);

  return { verified: true, reason: null, checks };
}

File: packages/go-verify/pq_hybrid.go
// SPDX-License-Identifier: Apache-2.0

// EP-HYBRID-v1 -- hybrid classical + post-quantum signature envelope for EP
// INFRASTRUCTURE keys (transparency-log, directory, checkpoint signing keys),
// ported from packages/verify/src/pq-hybrid.ts. Verification only.
//
// ENVELOPE
//
//	{
//	  "alg": "EP-HYBRID-v1",
//	  "signature_algos": ["Ed25519", "ML-DSA-65"],
//	  "sigs": { "Ed25519": "<base64url>", "ML-DSA-65": "<base64url>" }
//	}
//
// ANTI-STRIPPING (the property this file exists for). BOTH signatures are
// computed over a domain-separated signing input that INCLUDES a canonical
// encoding of signature_algos:
//
//	signing_input = UTF8("emilia-protocol/pq-hybrid/v1") || 0x00
//	             || UTF8(JSON(signature_algos))          || 0x00
//	             || message
//
// An attacker who strips the ML-DSA-65 signature and presents the Ed25519
// signature as a plain classical signature fails: the Ed25519 signature does
// not verify over the bare message, nor over a signing input committing to a
// reduced algorithm set. Removing an algorithm from the set changes what was
// signed, so every remaining signature fails. VerifyHybridEnvelope
// additionally requires the PRESENTED signature_algos to equal the registered
// set exactly (order-sensitive) and requires one signature per committed
// algorithm, no more and no fewer.
//
// FAIL-CLOSED. Missing, malformed, tampered, or reduced input refuses with a
// named reason. With no ML-DSA backend the result is a refusal with reason
// "pq_backend_unavailable": the PQ leg is never skipped and the envelope never
// verifies on the classical leg alone.
package emiliaverify

import "crypto/ed25519"

// HybridAlg is the envelope's alg marker.
const HybridAlg = "EP-HYBRID-v1"

// HybridSignatureAlgos is the registered algorithm set for EP-HYBRID-v1, in
// canonical order. v1 is a FIXED two-algorithm hybrid. Treat as read-only.
var HybridSignatureAlgos = []string{"Ed25519", "ML-DSA-65"}

// HybridDomain is the domain-separation label. The 0x00 separators keep label,
// algorithm set, and message unambiguous (JSON never contains a raw 0x00).
const HybridDomain = "emilia-protocol/pq-hybrid/v1"

// Named refusals. Byte-identical strings to HYBRID_REASONS in
// packages/verify/src/pq-hybrid.ts.
const (
	HybridReasonInvalidInput           = "invalid_input"
	HybridReasonInvalidEnvelope        = "invalid_envelope"
	HybridReasonAlgoSetMismatch        = "algo_set_mismatch"
	HybridReasonMissingSignature       = "missing_signature"
	HybridReasonMissingKey             = "missing_key"
	HybridReasonAlgorithmKeyMismatch   = "algorithm_key_mismatch"
	HybridReasonSignatureLengthInvalid = "signature_length_invalid"
	HybridReasonPublicKeyLengthInvalid = "public_key_length_invalid"
	HybridReasonClassicalInvalid       = "classical_signature_invalid"
	HybridReasonPQInvalid              = "pq_signature_invalid"
	HybridReasonPQBackendUnavailable   = "pq_backend_unavailable"
)

// HybridEnvelope is the EP-HYBRID-v1 wire shape.
type HybridEnvelope struct {
	Alg            string            `json:"alg"`
	SignatureAlgos []string          `json:"signature_algos"`
	Sigs           map[string]string `json:"sigs"`
}

// HybridVerificationKeys pins one key per leg. Ed25519PublicKey is base64url
// SPKI DER; MldsaPublicKey is base64url of the raw 1952-byte encoding.
type HybridVerificationKeys struct {
	Ed25519PublicKey string
	MldsaPublicKey   string
}

// HybridEnvelopeChecks records each independent step. A nil pointer means the
// step was never reached, which is distinct from false.
type HybridEnvelopeChecks struct {
	Envelope           bool  `json:"envelope"`
	AlgoSet            bool  `json:"algo_set"`
	ClassicalSignature *bool `json:"classical_signature"`
	PQSignature        *bool `json:"pq_signature"`
}

// HybridEnvelopeResult is the verdict. Reason is "" when the envelope
// verified.
type HybridEnvelopeResult struct {
	Verified bool                 `json:"verified"`
	Reason   string               `json:"reason"`
	Checks   HybridEnvelopeChecks `json:"checks"`
}

// HybridSigningInput builds the domain-separated signing input BOTH legs sign.
// It commits to the full signatureAlgos array; changing the array in any way
// (strip, reorder, substitute) changes these bytes. Returns nil when message
// is nil or signatureAlgos is empty or contains an empty string, mirroring the
// TypeError the JavaScript implementation raises for the same inputs.
func HybridSigningInput(message []byte, signatureAlgos []string) []byte {
	if message == nil || len(signatureAlgos) == 0 {
		return nil
	}
	algos := make([]any, 0, len(signatureAlgos))
	for _, a := range signatureAlgos {
		if a == "" {
			return nil
		}
		algos = append(algos, a)
	}
	out := make([]byte, 0, len(HybridDomain)+len(message)+64)
	out = append(out, HybridDomain...)
	out = append(out, 0x00)
	out = append(out, Canonicalize(algos)...)
	out = append(out, 0x00)
	out = append(out, message...)
	return out
}

func stringSlicesEqual(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// VerifyHybridEnvelope verifies an EP-HYBRID-v1 envelope. Verified is true
// only when ALL of the following hold:
//
//   - the envelope is well formed and carries alg "EP-HYBRID-v1";
//   - the presented signature_algos EXACTLY equals ["Ed25519","ML-DSA-65"],
//     order included (this is also what both signatures commit to, so
//     tampering fails twice);
//   - there is exactly one signature per committed algorithm, no extras;
//   - the Ed25519 signature is valid over the committed signing input, under a
//     CURVE-PINNED Ed25519 key and an exact 64-byte length pin;
//   - the ML-DSA-65 signature is valid over the committed signing input, under
//     a 1952-byte length-pinned key and an exact 3309-byte signature length
//     pin, checked by a REAL backend. With no backend the result is a refusal
//     with reason "pq_backend_unavailable"; the PQ leg is never skipped.
//
// It never panics on caller input. A nil message refuses with "invalid_input";
// a nil envelope refuses with "invalid_envelope".
func VerifyHybridEnvelope(message []byte, envelope *HybridEnvelope, keys *HybridVerificationKeys, opts AgilityOptions) HybridEnvelopeResult {
	checks := HybridEnvelopeChecks{}
	refuse := func(reason string) HybridEnvelopeResult {
		return HybridEnvelopeResult{Verified: false, Reason: reason, Checks: checks}
	}

	if message == nil {
		return refuse(HybridReasonInvalidInput)
	}

	// 1. Envelope shape (fail closed on anything unexpected).
	if envelope == nil {
		return refuse(HybridReasonInvalidEnvelope)
	}
	if envelope.Alg != HybridAlg {
		return refuse(HybridReasonInvalidEnvelope)
	}
	if envelope.Sigs == nil {
		return refuse(HybridReasonInvalidEnvelope)
	}
	checks.Envelope = true

	// 2. Algorithm-set commitment: the presented set must EXACTLY equal the
	//    registered set. Both signatures also commit to it cryptographically.
	if !stringSlicesEqual(envelope.SignatureAlgos, HybridSignatureAlgos) {
		return refuse(HybridReasonAlgoSetMismatch)
	}
	checks.AlgoSet = true

	// 3. Exactly one signature per committed algorithm; extras refuse.
	if len(envelope.Sigs) != len(HybridSignatureAlgos) {
		if len(envelope.Sigs) < len(HybridSignatureAlgos) {
			return refuse(HybridReasonMissingSignature)
		}
		return refuse(HybridReasonInvalidEnvelope)
	}
	for _, algo := range HybridSignatureAlgos {
		if envelope.Sigs[algo] == "" {
			return refuse(HybridReasonMissingSignature)
		}
	}

	// 4. Key material. The classical key is CURVE-PINNED to Ed25519 and the PQ
	//    key LENGTH-PINNED to the FIPS 204 parameter set, so neither leg can be
	//    verified under a substituted algorithm.
	if keys == nil || keys.Ed25519PublicKey == "" || keys.MldsaPublicKey == "" {
		return refuse(HybridReasonMissingKey)
	}
	edDER, err := b64urlDecode(keys.Ed25519PublicKey)
	if err != nil {
		return refuse(HybridReasonMissingKey)
	}
	edKey, mismatch := parsePinnedEd25519(edDER)
	if mismatch {
		return refuse(HybridReasonAlgorithmKeyMismatch)
	}
	if edKey == nil {
		return refuse(HybridReasonMissingKey)
	}
	pqKey, err := b64urlDecode(keys.MldsaPublicKey)
	if err != nil || len(pqKey) == 0 {
		return refuse(HybridReasonMissingKey)
	}
	if len(pqKey) != MLDSA65PublicKeyBytes {
		return refuse(HybridReasonPublicKeyLengthInvalid)
	}

	// Decode both signatures and pin their lengths to the declared algorithms
	// BEFORE any verify call. A relabeled Ed448 signature is 114 bytes, not 64;
	// the length pin is the second half of the anti-masquerade control and the
	// curve pin above is the first.
	edSig, err := b64urlDecode(envelope.Sigs["Ed25519"])
	if err != nil || len(edSig) != Ed25519SignatureBytes {
		return refuse(HybridReasonSignatureLengthInvalid)
	}
	pqSig, err := b64urlDecode(envelope.Sigs["ML-DSA-65"])
	if err != nil || len(pqSig) != MLDSA65SignatureBytes {
		return refuse(HybridReasonSignatureLengthInvalid)
	}

	signingInput := HybridSigningInput(message, envelope.SignatureAlgos)
	if signingInput == nil {
		return refuse(HybridReasonInvalidInput)
	}

	// 5. Classical leg (Ed25519) over the committed signing input.
	edOK := ed25519.Verify(edKey, signingInput, edSig)
	checks.ClassicalSignature = boolPtr(edOK)
	if !edOK {
		return refuse(HybridReasonClassicalInvalid)
	}

	// 6. PQ leg (ML-DSA-65). No backend means REFUSE. Never skip, never pass.
	if opts.Mldsa == nil {
		return refuse(HybridReasonPQBackendUnavailable)
	}
	pqOK := opts.Mldsa.Verify(pqSig, signingInput, pqKey)
	checks.PQSignature = boolPtr(pqOK)
	if !pqOK {
		return refuse(HybridReasonPQInvalid)
	}

	return HybridEnvelopeResult{Verified: true, Reason: "", Checks: checks}
}
